action/protocol/rewarding/fund.go

```diff
@@ -119,6 +119,73 @@ func (p *Protocol) deposit(
 	}, nil
 }
 
+func (p *Protocol) depositAndBurn(
+	ctx context.Context,
+	sm protocol.StateManager,
+	burnAddr address.Address,
+	amount, burnAmount *big.Int,
+	transactionLogType iotextypes.TransactionLogType,
+) ([]*action.TransactionLog, error) {
+	var (
+		actionCtx           = protocol.MustGetActionCtx(ctx)
+		fCtx                = protocol.MustGetFeatureCtx(ctx)
+		accountCreationOpts = []state.AccountCreationOption{}
+	)
+	if fCtx.CreateLegacyNonceAccount {
+		accountCreationOpts = append(accountCreationOpts, state.LegacyNonceAccountTypeOption())
+	}
+	// Subtract balance from caller
+	acc, err := accountutil.LoadAccount(sm, actionCtx.Caller, accountCreationOpts...)
+	if err != nil {
+		return nil, err
+	}
+	if err := acc.SubBalance(amount); err != nil {
+		return nil, err
+	}
+	if !isZero(burnAmount) {
+		if err := acc.SubBalance(burnAmount); err != nil {
+			return nil, err
+		}
+	}
+	if err := accountutil.StoreAccount(sm, actionCtx.Caller, acc); err != nil {
+		return nil, err
+	}
+	// Add balance to fund
+	var (
+		f    = fund{}
+		tLog = []*action.TransactionLog{
+			&action.TransactionLog{
+				Type:      transactionLogType,
+				Sender:    actionCtx.Caller.String(),
+				Recipient: address.RewardingPoolAddr,
+				Amount:    amount,
+			},
+		}
+	)
+	if _, err := p.state(ctx, sm, _fundKey, &f); err != nil {
+		return nil, err
+	}
+	f.totalBalance.Add(f.totalBalance, amount)
+	f.unclaimedBalance.Add(f.unclaimedBalance, amount)
+	if !isZero(burnAmount) {
+		f.unclaimedBalance.Sub(f.unclaimedBalance, burnAmount)
+		if f.unclaimedBalance.Sign() == -1 {
+			return nil, errors.New("no enough available balance")
+		}
+		// TODO: add burnAmount to burnAddr
+		tLog = append(tLog, &action.TransactionLog{
+			Type:      transactionLogType,
+			Sender:    actionCtx.Caller.String(),
+			Recipient: burnAddr.String(),
+			Amount:    burnAmount,
```

Comment:
according to the transaction logs, the `burnAmount` and the `amount` are both from `sender`. Somehow, the `burnAmount` is subtracted from the `unclaimedBalance`. Question: which one is the direct source of `burnAmount`, pool or sender?

Reply:
good catch, it should not be subtracted again (since `amount` is not the total, does not include `burnAmount`)